### ISSUE TITLE
6to5-node breaks gulp watch

### ISSUE THREAD

**Issue by ivan-kleshnin** [6to5/6to5#489]
*Disclaimer: I'm not quite sure it's a problem exactly with 6to5 and not gulp.*

**gulp watch  task does not work with 6to5-node** or vice-versa **6to5-node breaks gulp watch**

I change shebang in `/usr/local/lib/node_modules/gulp/bin/gulp.js` file
```
#!/usr/bin/env node --> 
#!/usr/bin/env 6to5-node
```

Then I try to run any gulp watch task

```
gulp.task("watch", function() {
  // It does not matter what task is – every one I tried crashed.
  gulp.watch("./src/app_react/**/*.js", ["dist-react"]);
});
```

Then it crashes in byzantine way

```
TypeError: boolean is not a function
    at Function.isPlainObject (/Users/ivankleshnin/JavaScript/twitto/node_modules/gulp/node_modules/vinyl-fs/node_modules/glob-watcher/node_modules/gaze/node_modules/globule/node_modules/lodash/dist/lodash.js:1652:64)
    at Object.globule.find (/Users/ivankleshnin/JavaScript/twitto/node_modules/gulp/node_modules/vinyl-fs/node_modules/glob-watcher/node_modules/gaze/node_modules/globule/lib/globule.js:64:19)
    at Gaze.add (/Users/ivankleshnin/JavaScript/twitto/node_modules/gulp/node_modules/vinyl-fs/node_modules/glob-watcher/node_modules/gaze/lib/gaze.js:174:19)
    at new Gaze (/Users/ivankleshnin/JavaScript/twitto/node_modules/gulp/node_modules/vinyl-fs/node_modules/glob-watcher/node_modules/gaze/lib/gaze.js:74:10)
    at gaze (/Users/ivankleshnin/JavaScript/twitto/node_modules/gulp/node_modules/vinyl-fs/node_modules/glob-watcher/node_modules/gaze/lib/gaze.js:86:10)
    at Object.module.exports [as watch] (/Users/ivankleshnin/JavaScript/twitto/node_modules/gulp/node_modules/vinyl-fs/node_modules/glob-watcher/index.js:12:17)
    at Gulp.watch (/Users/ivankleshnin/JavaScript/twitto/node_modules/gulp/index.js:35:16)
    at Gulp.<anonymous> (/Users/ivankleshnin/JavaScript/twitto/gulpfile.js:78:8)
    at module.exports (/Users/ivankleshnin/JavaScript/twitto/node_modules/gulp/node_modules/orchestrator/lib/runTask.js:34:7)
    at Gulp.Orchestrator._runTask (/Users/ivankleshnin/JavaScript/twitto/node_modules/gulp/node_modules/orchestrator/index.js:273:3)
Error running task sequence: { task: 'watch',
  message: 'watch catch',
  duration: 0.001110806,
  hrDuration: [ 0, 1110806 ],
  err: [TypeError: boolean is not a function] }
```

Other gulp things except **gulp.watch** seems to work ok.

If you think it's a Gulp problem – I'll repost it there.

**Comment by sebmck** [comment]
Looks like it's a result of gulp depending on [vinyl-fs](https://github.com/wearefractal/vinyl-fs) that depends on [glob-watcher](https://github.com/wearefractal/glob-watcher) that uses an outdated version of [gaze](https://github.com/shama/gaze) which uses an outdated version of [globule](https://github.com/cowboy/node-globule) that uses an oudated version of [lodash](https://github.com/lodash/lodash) that has a dodgy `Object.getPrototypeOf` check because [core-js](https://github.com/zloirock/core-js) overrides it.

One possible fix (and I have no idea if this will work) is to add lodash to your dependencies via `npm install --save lodash` and then reinstall all of your `node_modules`.

**Comment by ivan-kleshnin** [comment]
sebmck is a genius :alien: .  I installed latest **lodash** as project dependency and removed it from **globule** and it works now. Submitted related issue to **glob-watcher**

**Comment by jdalton** [comment]
>  of lodash that has a dodgy Object.getPrototypeOf check because core-js overrides it.

lodash correctly detects `_.isNative(Object.getPrototypeOf)` as `false`.
See http://jsbin.com/pugimomame/1/.

**Comment by sebmck** [comment]
@jdalton Yeah, I was referring to the version that was being loaded by `globule`. It was using version 1.0.1 which has the following:

```javascript
var isPlainObject = function(value) {
  if (!(value && typeof value == 'object')) {
    return false;
  }
  var valueOf = value.valueOf,
      objProto = typeof valueOf == 'function' && (objProto = getPrototypeOf(valueOf)) && getPrototypeOf(objProto);

  return objProto
    ? value == objProto || (getPrototypeOf(value) == objProto && !isArguments(value))
    : shimIsPlainObject(value);
};
```

It was calling `getPrototypeOf` which was set via:

```javascript
getPrototypeOf = reNative.test(getPrototypeOf = Object.getPrototypeOf) && getPrototypeOf
```

`getPrototypeOf` however was `false` as `Object.getPrototypeOf` was shimmed via [core-js](https://github.com/zloirock/core-js) and was failing the `reNative` test.

**Comment by sebmck** [comment]
`globule` has version `2.4.1` listed in it's dependencies though so I'm not sure why npm is using `1.0.1`.

You can reproduce this with the following steps:

```sh
$ npm install gulp core-js
```

```javascript
> require("core-js");
> require("gulp/node_modules/vinyl-fs/node_modules/glob-watcher/node_modules/gaze/node_modules/globule/node_modules/lodash").isPlainObject([]);
```

**Comment by jdalton** [comment]
Ah nice, so just an old bug. Thank you for digging.

**Comment by gsklee** [comment]
I'm not sure if this is an old bug because I'm getting this in the latest versions of things (babel@4.0.2, gulp@3.8.11), or there might be a regression of sorts.

**Comment by sebmck** [comment]
It's an old bug in lodash, see my comment [here](https://github.com/babel/babel/issues/489#issuecomment-69919417) that has since been fixed. It's the result of a chain of outdated dependencies.

**Comment by gsklee** [comment]
@sebmck So if I'm getting this right, I'll need to ask `glob-watcher` to use a newer version of `gaze`?

**Comment by jdalton** [comment]
> It's an old bug in lodash,

Yap a bug in lodash 1.0.1 (2 years old)

**Comment by qrohlf** [comment]
So to clarify for others who are running into this, the workaround is to `npm install -S loadash` to grab the latest lodash as a project dependency, then edit

```
node_modules/gulp/node_modules/vinyl-fs/node_modules/glob-watcher/node_modules/gaze/node_modules/globule
```

to remove the `"lodash": "~1.0.1",` dependency, then remove 

```
node_modules/gulp/node_modules/vinyl-fs/node_modules/glob-watcher/node_modules/gaze/node_modules/globule/node_modules/lodash
```

then things should work...


![enough-internet](https://cloud.githubusercontent.com/assets/347189/6364109/9a0b5926-bc56-11e4-8616-86e76058986f.gif)

**Comment by jdalton** [comment]
Naw, until vinyl-fs bumps you can reference it's package via its commit hash:
```json
"vinyl-fs": "git://github.com/wearefractal/vinyl-fs.git#6d7952973d026246602d59f1358d806463fe508f"
```

**Comment by jdalton** [comment]
hmm ah ya because of `gulp` hmm \cc @contra

**Comment by contra** [comment]
Yeah we haven't published the 4.0 version of vinyl-fs yet because we have a few breaking PRs we want to land first. Shouldn't be long though, until then just use the master branch or something in your package

**Comment by AgentME** [comment]
I came up with a work-around that doesn't require everyone using a project to mess with their node_modules folder:

```javascript
require('babel/register');
Object.getPrototypeOf.toString = function() {return Object.toString();};
```

glob-watcher isn't updating their gaze dependency, so it seems like one of the following has to happen to get this properly fixed:

* globule backports the lodash dependency update and releases a version 0.1.1.
* gaze backports the globule dependency update and releases a version 0.5.2.
* lodash backports the getPrototypeOf fix and releases a version 1.0.2.
* babel changes its polyfill to make its Object.getPrototypeOf replacement look native to work around the ancient lodash issue.

**Comment by jdalton** [comment]
> lodash backports the getPrototypeOf fix and releases a version 1.0.2.

I can do that.

**Comment by AgentME** [comment]
@jdalton That'd be perfect, thanks.

**Comment by jdalton** [comment]
Published a fix `lodash@1.0.2`.

**Comment by gsklee** [comment]
@jdalton :+1:
